math/VectorArithmetic.cpp

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) 2020 Nicholas Corgan
+// SPDX-License-Identifier: BSL-1.0
+
+#include <Pothos/Callable.hpp>
+#include <Pothos/Exception.hpp>
+#include <Pothos/Framework.hpp>
+#include <Pothos/Object.hpp>
+
+#include <complex>
+#include <cstdint>
+#include <string>
+#include <vector>
+
+//
+// Arithmetic functions
+//
+
+template <typename T>
+void XPlusK(const T* in, const std::vector<T>& kVec, T* out, size_t len)
+{
+    for(size_t i = 0; i < len; ++i)
+    {
+        out[i] = in[i] + kVec[i % kVec.size()];
+    }
+}
+
+template <typename T>
+void XSubK(const T* in, const std::vector<T>& kVec, T* out, size_t len)
+{
+    for(size_t i = 0; i < len; ++i)
+    {
+        out[i] = in[i] - kVec[i % kVec.size()];
+    }
+}
+
+template <typename T>
+void KSubX(const T* in, const std::vector<T>& kVec, T* out, size_t len)
+{
+    for(size_t i = 0; i < len; ++i)
+    {
+        out[i] = kVec[i % kVec.size()] - in[i];
+    }
+}
+
+template <typename T>
+void XMultK(const T* in, const std::vector<T>& kVec, T* out, size_t len)
+{
+    for(size_t i = 0; i < len; ++i)
+    {
+        out[i] = in[i] * kVec[i % kVec.size()];
+    }
+}
+
+template <typename T>
+void XDivK(const T* in, const std::vector<T>& kVec, T* out, size_t len)
+{
+    for(size_t i = 0; i < len; ++i)
+    {
+        out[i] = in[i] / kVec[i % kVec.size()];
+    }
+}
+
+template <typename T>
+void KDivX(const T* in, const std::vector<T>& kVec, T* out, size_t len)
+{
+    for(size_t i = 0; i < len; ++i)
+    {
+        out[i] = kVec[i % kVec.size()] / in[i];
+    }
+}
+
+//
+// Test class
+//
+
+template <typename T, void (*Operator)(const T*, const std::vector<T>&, T*, size_t)>
+class VectorArithmetic: public Pothos::Block
+{
+public:
+    using Class = VectorArithmetic<T, Operator>;
+
+    VectorArithmetic(size_t dimension):
+       Pothos::Block(),
+       _vector({0})
+    {
+        const Pothos::DType dtype(typeid(T), dimension);
```

Comment:
Not taken into account. Each element of the vector should apply to all elements in one instance of the DType.